examples/nemo/eg1/runme_openmp_gpu.py

```diff
@@ -0,0 +1,102 @@
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: S. Siso, STFC Daresbury Lab
+
+'''A simple test script showing the introduction of OpenMP for GPU with
+PSyclone. In order to use it you must first install PSyclone. See README.md
+in the top-level psyclone directory.
+
+Once you have psyclone installed, this script may be run by doing:
+
+ >>> python runme_openmp_gpu.py
+
+This should generate the Fortran with OpenMP target and loop directives.
+'''
+
+from __future__ import print_function
+from psyclone.parse.algorithm import parse
+from psyclone.psyGen import PSyFactory, TransInfo
+from psyclone.psyir.nodes import Loop, Assignment, CodeBlock
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans
+
+API = "nemo"
+USE_GPU = True  # Enable for generating OpenMP target directives
+
+
+def main():
+    ''' Parser the Nemo tra_adv file and transform it to OpenMP for GPU '''
+    _, invokeinfo = parse("../code/tra_adv.F90", api=API)
+    psy = PSyFactory(API).create(invokeinfo)
+    sched = psy.invokes.get('tra_adv').schedule
+
+    omp_loop_trans = TransInfo().get_trans_name('OMPLoopTrans')
+    omp_loop_trans.omp_worksharing = False
+    omp_target_trans = TransInfo().get_trans_name('OMPTargetTrans')
+
+    # Convert all array implicit loops to explicit loops
+    explicit_loops = NemoAllArrayRange2LoopTrans()
+    for assignment in sched.walk(Assignment):
+        try:
+            explicit_loops.apply(assignment)
+        except KeyError:
+            # The transformation just works if jpi and jpj are
```

Comment:
s/just/only/

Reply:
I think you also need to filter which Assignments to transform by using the `is_array_range` property. Either that or extend the try/catch so that it handles TransformationErrors too. That way this script will be more general purpose (useful as users may take it as a starting point).

Reply:
Removed the exception as it doesn't happen in traadv.f90 (and I am reworking this in another branch)

Regarding your second comment note that this transformation only produces TransformationError if it is applied to the wrong (non-assignment) node. The transformation itself recurses as much as it can over applying another transformation and ignores the TransformationErrors that this second one produced. I believe this already filters (the inner transformation raises and the outer ignores) the array without ranges issue that you mention.

Reply:
Aha - thanks for explaining.